Fix for engineered modules in SLEF imports from Inara.

alexeygorb opened the pull request:
A final approach at #631 which fixes the inconsistency issues between the Inara and EDSM version of imports.

alexeygorb commented:
Hi, @felixlinker could you please take a look at this PR? :)

felixlinker commented:
Okay, I'm starting to become a bit skeptical! The design overall looks good. But I fear, the resistance handling might be duplicated with other code now. Don't get me wrong - this is mostly coriolis's fault given its current state. The whole design doesn't allow to call parts of the same logic in different places. But we shouldn't worsen things. Once I find time, which unfortunately might take a while, I will have another look.

alexeygorb commented:
@felixlinker 
Thanks for the feedback, and I get the point about the code duplication and I totally agree. :)
I'll try to separate the resistance calculation into a function, similar to what I did with module quality when I get some spare time

alexeygorb commented:
Hi, @felixlinker 
I went a little bit extra and did a small refactoring on how the modifications are applied.
There are 2 main changes:
- I've removed the 'sentFromUI' flag, since it always duplicates the 'isAbsolute' value.
- and I've moved the resistances calculations/adjustments inside the setModValue method, cause in the test cases it's only adjust when the value is not 'absolute'

Please let me know what you think :)

felixlinker commented:
Changes look really good upon first glance! I have to look at them more closely, tough. I hope I find time next week! Am with my family right now.

alexeygorb commented on src/app/utils/BlueprintFunctions.js:
```diff
@@ -327,26 +327,43 @@ export function setPercent(ship, m, percent) {
+ * @param {Number}   quality    The quality to apply - float number 0 to 1.
+ * @param {Function} cb         The Callback to run for each property. Function (featureName, value)
+ */
+export function setQualityCB(blueprint, quality, cb) {
```
This call back is called for each modified module property, since one blueprint has multiple properties and thus the call back will be fired multiple times.
Alternatively I can change this function to return a set of properties and then they can be cycled from where the function is called.

felixlinker replied:
Not necessary. In my mental model, callbacks are only used for asynchronous behavior. But you're right, e.g., `forEach` also takes callback.

alexeygorb commented on src/app/utils/JournalUtils.js:
```diff
@@ -295,28 +293,10 @@ function _addModifications(module, modifiers, quality, blueprint, grade, special
       // Carry out the required changes
       for (const action in modifierActions) {
-        if (isNaN(modifierActions[action])) {
-          module.setModValue(action, modifierActions[action]);
```
`modifierActions` is an element from coriolis-data, I checked all the possible actions in `modifierActions.json` and there is always a number. So it looks like this branch would never get triggered.

felixlinker replied:
Same here as with the other dropped branch. `damagedist` can be an object, and there `isNaN` evaluates to `true`. So this branch needs to be restored as well!

alexeygorb replied:
Well, I can restore the branch to reduce risk of breaking anything :)
But if you check the Module.setModValue function the operations taken on the second parameter (value) are purely mathematical: `value = value - modifierActions[name];` or `value = ((value / 10000 + 1) / (1 + modifierActions[name]) - 1) * 10000;` so passing an object would just set the value to NaNs...
Of course I might just not be getting it, so I will restore the branch I deleted anyway to be 100% sure

felixlinker replied:
This calculations, though, are only triggered if `name` is some resistance, or `valueiswithspecial` is true. In the `isNaN` branch `valueiswithspecial` is undefined so just the object (in this case it *is* an object) will be set.

alexeygorb commented on src/app/utils/BlueprintFunctions.js:
```diff
@@ -360,23 +377,6 @@ export function setRandom(ship, m) {
-  } else if (Modifications.modifications[featureName].type == 'numeric') {
-    ship.setModification(m, featureName, value * 100);
-  } else {
-    ship.setModification(m, featureName, value);
```
Same as with modifierActions in another example, there is always a numeric value in the respective coriolis-data file. And then, it doesn't make much sense to set the value if it's a string or an object, then why risk and pass the unknown value to the module modification. Maybe it would make sense to log it instead?

felixlinker replied:
Note that `type` here is not the data type. This can also be `object`, i.e., for `damagedist`. So the branch needs to be restored!

felixlinker replied:
Note that `type` here is not the data type. This can also be `object`, i.e., for `damagedist`. So the branch needs to be restored!

alexeygorb replied:
@felixlinker I just checked  - this branch was not dropped. This is how this part looks now:
```
if (Modifications.modifications[featureName].type == 'percentage') {
      value = value * 10000;
    } else if (Modifications.modifications[featureName].type == 'numeric') {
      value = value * 100;
    }

    cb(featureName, value);
```
So, it was removed, but the logic remains - the value will be passed as is, this else would just be redundant then.

felixlinker commented:
Thanks for bearing with me! The coriolis codebase is incredibly confusing. I find you did a stellar job in digging through it and attempting to simplify it! Thank you so much for these changes!

felixlinker commented:
Changes are live on https://beta.coriolis.io I will try to merge them end of next week, if I don't hear about issues :)

alexeygorb commented:
Thank you for accepting and merging! And I'm really glad that I could help. :)

designerzim commented:
Thanks for the fixes! I tried my whole fleet with mostly non-maxed engineering. The `import?data` comes through just fine now.

felixlinker commented:
Changes live on https://coriolis.io !
